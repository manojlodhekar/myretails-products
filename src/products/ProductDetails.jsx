import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Image, Item, Icon, Grid, Segment} from 'semantic-ui-react';
import PropTypes from 'prop-types';
import { getProductDetails } from '../actions/getProductDetails';
import ProductHeader from './ProductHeader';
import productConst from '../const/productConst';
import ProductPromotions from './ProductPromotions';
import ProductReturnPolicy from './ProductReturnPolicy'
import HTMLRenderer from './HTMLRenderer';
import ProductReview from './ProductReview';
import ProductCarousel from './ProductCarousel';
import ProductsZoomDialog from './ProductsZoomDialog';

class ProductDetails extends Component {
    
    constructor(props){
        super(props);
        this.state = {
            quantity : 1,
            windowWidth:window.innerWidth,
            currentActiveIndex: 0,
            open: false
        }
        this.updateDimensions = this.updateDimensions.bind(this);
        this.buyProduct = this.buyProduct.bind(this);
    };

    componentDidMount(){
        this.props.actions.getProductDetails();
        window.addEventListener("resize", this.updateDimensions);
    }

    buyProduct = (btnStatus) =>{
        alert(` ${this.state.quantity} items added to cart.`);
    }
    updateDimensions() {
        const realContainerWidth = window.innerWidth;
        this.setState({
            windowWidth:realContainerWidth 
        })
      }
    changeQuantity = (operation) => {
        if(operation === 'remove'){
            if(this.state.quantity === 1)
            return;
            this.setState({ quantity: this.state.quantity - 1 })
        }
        else{
            this.setState({ quantity: this.state.quantity + 1})
        }
      }

      setHtmlContent = (content) =>{
          if(!content)return;

          let productDesc = content.map(item => {
            return(`<li>${item}</li>`)
          });

          return productDesc;
      }

      selectRetailsProductImage = (index , url) =>{
        console.log('index = ' + index + " URL" + url);
        this.setState({currentActiveIndex: index})
      }
      openZoomedImage = () =>{
        this.setState({ open: true })
      }
      modalPopupClosed = () =>{
        this.setState({
          open: false
        })
      }
      getCurrentImageUrl = () =>{
          if(this.props.productDetailImages.AlternateImages){
              return this.props.productDetailImages.AlternateImages;
            }
          
      }

    render(){

        const {selectedProductDetail, productDetailImages, productDetailDescription, productDetailCustomerReview, productDetailOffers,
            productDetailPromotions, productDetailReturnPolicy} = this.props;
            
        let productsDesc =  this.setHtmlContent(productDetailDescription.features);
        productsDesc = productsDesc ? `<ul> ${productsDesc.join('')} </ul>` : '';

        if(selectedProductDetail.CatalogEntryView && productDetailImages.PrimaryImage  && !this.imageCounter){
            this.imageCounter = !this.imageCounter;
            productDetailImages.AlternateImages.unshift(productDetailImages.PrimaryImage[0]);
        }
        
        return(
            
                <div className="detailsContainer">
                    <ProductHeader subHeaderText={productConst.PRODUCT_MAIN_PAGE_SUBHEADER} />
                {
                    selectedProductDetail.itemId ?
                    <div>
                        <Item.Group>
                            <Item>
                            <div className="productLeft">
   
                                <div className="productTitle">{selectedProductDetail.title}</div> 
                                <Item.Image size='large' src={productDetailImages.AlternateImages[this.state.currentActiveIndex].image}
                                style={{paddingLeft: '10%'}}/>
                                <div className="viewLargeIcon" onClick={(e) => this.openZoomedImage()}>
                                    <Icon name='zoom' /> View Larger
                                </div>
                                <ProductCarousel productsList={productDetailImages.AlternateImages} currentActiveIndex={this.state.currentActiveIndex} 
                                selectProductImage = {this.selectRetailsProductImage} />
                                {
                                    this.state.windowWidth >= 768 ?
                                    <ProductReview ProductReviewDetail={productDetailCustomerReview} />:
                                    null
                                }
                            
                            </div>
                        <Item.Content className='productContent'>
                        <Item.Meta className='productPrice'>{productDetailOffers.OfferPrice[0].formattedPriceValue} <span className="offerPriceTxt"> {productDetailOffers.OfferPrice[0].priceQualifier}</span> </Item.Meta>
                            <ProductPromotions promotionDetails={productDetailPromotions} />
                            <div className="quantityCls">
                                <span>Quantity: </span>  <span className="quantityIconCls"> <Icon name="minus circle" onClick={(e) => this.changeQuantity('remove')}/> <span className="category">{this.state.quantity} </span> <Icon name="plus circle" onClick={(e) =>this.changeQuantity('add')}/> </span>
                            </div>
                                {
                                (selectedProductDetail.inventoryStatus === 'Online' && (selectedProductDetail.purchasingChannelCode === '0' || selectedProductDetail.purchasingChannelCode === '2')) ?
                                    <Image src={productConst.ADD_TO_CART} className="buttonCls" onClick={ (e) => this.buyProduct(productConst.CART_BTN_LABEL)} /> 
                                    :
                                    <div>
                                        <Image src={productConst.PICK_IN_STORE} className="buttonCls" onClick={(e) => this.buyProduct(productConst.PICK_BTN_LABEL)} />
                                        <span style={{padding: 50}}>{productConst.FIND_STORE} </span>
                                    </div>
                                }

                            <ProductReturnPolicy returnPolicyDetails={productDetailReturnPolicy} />
                            <div style={{width: '100%', paddingLeft: 25, paddingBottom: 15}}>

                            <Grid columns={3}>
                                <Grid.Row>
                                <Grid.Column className="segCol">
                                     <Segment className="segmentItem" onClick={()=>alert(`Click to action for "${productConst.ADD_TO_REG}"`)}>{productConst.ADD_TO_REG}</Segment>
                                </Grid.Column>
                                <Grid.Column className="segCol">
                                     <Segment className="segmentItem" onClick={()=>alert(`Click to action for "${productConst.ADD_TO_LIST}"`)}>{productConst.ADD_TO_LIST}</Segment>
                                </Grid.Column>
                                <Grid.Column className="segCol">
                                     <Segment className="segmentItem" onClick={()=>alert(`Click to action for "${productConst.SHARE}"`)}>{productConst.SHARE}</Segment>
                                </Grid.Column>
                                </Grid.Row>
                            </Grid>

                            </div>
                            <div>
                                <span className="productTitle">{productConst.PRODUCT_HIGHLIGHT}</span>
                                    <HTMLRenderer htmlContents={productsDesc} />
                            </div>
                            {
                            this.state.windowWidth < 768 ?
                            <ProductReview ProductReviewDetail={productDetailCustomerReview} />:
                            null
                            }

                            
                            </Item.Content>
                        </Item>
                    </Item.Group>
                    <ProductsZoomDialog openPopup={this.state.open}
        imageURLs={this.getCurrentImageUrl()}
        popupClosed = {this.modalPopupClosed}
        currentImgIndex={this.state.currentActiveIndex}/>
                    </div>
                    :
                    <div> Loading....</div>
                }
                
        </div>
        );
    };
}

ProductDetails.propTypes = {
    selectedProductDetail: PropTypes.object.isRequired,
    productDetailImages: PropTypes.object.isRequired,
    productDetailDescription: PropTypes.object.isRequired,
    productDetailCustomerReview: PropTypes.object.isRequired,
    productDetailOffers: PropTypes.object.isRequired,
    productDetailPromotions: PropTypes.array.isRequired,
    productDetailReturnPolicy: PropTypes.object.isRequired,
    actions: PropTypes.object.isRequired
}
function mapStateToProps(state){
    return{
        selectedProductDetail: state.products.selectedProductDetail,
        productDetailImages: state.products.productDetailImages,
        productDetailDescription: state.products.productDetailDescription,
        productDetailCustomerReview: state.products.productDetailCustomerReview,
        productDetailOffers: state.products.productDetailOffers,
        productDetailPromotions: state.products.productDetailPromotions,
        productDetailReturnPolicy: state.products.productDetailReturnPolicy
    };
}

function mapDispatchToProps(dispatch){
    return {
        actions: bindActionCreators( { getProductDetails }, dispatch),
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(ProductDetails);
