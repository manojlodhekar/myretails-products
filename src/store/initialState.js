export default {
    products: {
        selectedProductDetail: {},
        productDetailImages:{},
        productDetailDescription:{},
        productDetailCustomerReview:{},
        productDetailOffers:{},
        productDetailPromotions:[],
        productDetailReturnPolicy:{},
        productApiError: false
    }
}