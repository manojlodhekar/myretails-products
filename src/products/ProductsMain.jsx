import React, { Component } from 'react';
import Snackbar from '@material-ui/core/Snackbar';
import productConst from '../const/productConst';
import { connect } from 'react-redux';
import {setApiError} from '../actions/getProductDetails';
import { bindActionCreators } from 'redux';
import ProductDetails from './ProductDetails';

/** This is a description of the foo function. */
class ProductsMain extends Component {
/**
 * Represents a book.
 * @constructor
 */
    constructor(props){
        super(props);
        this.state = {
            open : false
        }
    };


 /**
 * Represents a book.
 * @constructor
 * @param {string} title - The title of the book.
 * @param {string} author - The author of the book.
 */
    handleRequestClose = () => {
        this.props.actions.setApiError(false)
      };
    render(){
        const {productApiError} = this.props;
        return(
           <div>
               <ProductDetails />
               <Snackbar
                    className="snakeBarCls"
                    open={productApiError}
                    message={productConst.ERROR_MESSAGE}
                    action="undo"
                    autoHideDuration={3000}
                    onClose={this.handleRequestClose}
                />
            </div>
        );
    };
}

function mapStateToProps(state){
    return{
        productApiError: state.products.productApiError
    };
}

function mapDispatchToProps(dispatch){
    return {
        actions: bindActionCreators( { setApiError }, dispatch),
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(ProductsMain);