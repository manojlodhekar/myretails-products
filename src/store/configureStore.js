import { createStore, applyMiddleware } from 'redux';
import rootReducer from '../reducers';
import thunk from 'redux-thunk';    
import initialState from './initialState'

export default function configureStore() {
    return (createStore(
        rootReducer,initialState,applyMiddleware(thunk)
    ));
}