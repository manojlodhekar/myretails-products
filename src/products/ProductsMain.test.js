import React from 'react';
import ProductsMain from './ProductsMain';
import configureStore from 'redux-mock-store';
import ProductDetails from './ProductDetails';
import Snackbar from '@material-ui/core/Snackbar';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, mount } from 'enzyme';
import initialState from '../store/initialState'

jest.mock('react-dom');

Enzyme.configure({ adapter: new Adapter() });
const mockStore = configureStore();
const store = mockStore(initialState);
const productsMain = mount(
    <ProductsMain store={store}>
      <ProductDetails />
      <Snackbar />
    </ProductsMain>
);

describe('ProductsMain is defined', () => {

  it('should be defined', () => {
    expect(productsMain).toBeDefined();
  });
  it('should render correctly', () => {
    expect(productsMain).toHaveLength(1);
  });

  it('should render inner component', () => {
    expect(productsMain).toMatchSnapshot();
  });

  it('should set default open state component', () => {
    const productsMain = shallow(
      <ProductsMain store={store} />
  );
  productsMain.setState({open: true});
    let openState = productsMain.state().open
    expect(openState).toEqual(true);
  });

 });
