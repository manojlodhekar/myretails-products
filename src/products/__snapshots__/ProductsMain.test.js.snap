// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsMain is defined should render inner component 1`] = `
ReactWrapper {
  "length": 1,
  Symbol(enzyme.__unrendered__): <Connect(ProductsMain)
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
>
    <Connect(ProductDetails) />
    <WithStyles(Snackbar) />
</Connect(ProductsMain)>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
